### PR TITLE
[Golang] Add support allOf inheritance with embedded go structs

### PR DESCRIPTION
**Pull request by sebastien-rosset**
This PR addresses the following:
* The generated go-experimental code is not compiling in some cases.
* Use an embedded struct to model the codegen parent
* The instantiation type of a parent of "map" type is incorrect

In particular, the generated samples/openapi3/client/petstore/go-experimental/go-petstore/api_fake.go file has the following code:
```
 t := *r.pipe
  if reflect.TypeOf(t).Kind() == reflect.Slice {
     ....
  }
  ....
  t := *r.context
```

As you can see, the same variable **"t"** is defined twice, which causes a compilation error:
```
go-petstore/api_fake.go:1782:4: no new variables on left side of :=
```

This PR depends on #5075.

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### CODE REVIEW

**Comment by sebastien-rosset** [comment]
@antihax @bvwells @grokify @kemokemo @bkabrda

**Comment by sebastien-rosset** [comment]
It looks like "continuous-integration/travis-ci/pr " is failing for a reason other than this PR.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractGoCodegen.java [hunk 1]
```diff
@@ -328,6 +329,16 @@ public String getSchemaType(Schema p) {
             type = openAPIType;
         return type;
     }
+    
+    @Override
+    public String toInstantiationType(Schema property) {
```
Comment (bkabrda) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractGoCodegen.java]: Could you please provide a comment here on what this method does and why it does that? I *think* I understand it in the context of your PR, but if I was just looking at it in the codebase, I'd probably have no idea what's happening here and why.

File: modules/openapi-generator/src/main/resources/go-experimental/api.mustache [hunk 2]
```diff
@@ -134,14 +134,16 @@ func (a *{{{classname}}}Service) {{{nickname}}}(ctx _context.Context{{#hasParams
 	{{#queryParams}}
 	{{#required}}
 	{{#isCollectionFormatMulti}}
-	t:={{paramName}}
-	if reflect.TypeOf(t).Kind() == reflect.Slice {
-		s := reflect.ValueOf(t)
-		for i := 0; i < s.Len(); i++ {
-			localVarQueryParams.Add("{{baseName}}", parameterToString(s.Index(i), "{{#collectionFormat}}{{collectionFormat}}{{/collectionFormat}}"))
+	{
```
Comment (bkabrda) [inline modules/openapi-generator/src/main/resources/go-experimental/api.mustache]: Why is this change necessary? Is it to make sure that the scope of variables defined in this block is limited?

**Comment by sebastien-rosset** [comment]
> So after the explanation provided in [#4607 (comment)](https://github.com/OpenAPITools/openapi-generator/pull/4607#issuecomment-566662587) I think I'm ok with accepting this change. It seems to work ok and makes sense to me. I left one suggestion to improve and one question inline to get some improvement/explanation.
> 
> What I'm not sure about is whether this should go inside the non-experimental go client, as I think it might result in unexpected changes in JSON payloads.

Using Golang embedded structs have zero impact on the JSON payload (vs a flattened structure). The payload remains exactly the same.

> OTOH this feature never worked before, so I guess adding it can also be considered just an improvement/fix and the different JSON payloads will now actually be correct.
> 
> All in all, I'm ok with this being merged (after the suggested change/reply), assuming it gets a 👍 from someone else as well (@antihax WDYT?)

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractGoCodegen.java (hunk 1 above)
Comment (sebastien-rosset) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractGoCodegen.java]: sure, will do.

**Comment by bkabrda** [comment]
> Using Golang embedded structs have zero impact on the JSON payload (vs a flattened structure). The payload remains exactly the same.

Yeah, but the serialized children will now contain parent attributes, which they didn't before. So the payload is going to be different, is it not?

File: modules/openapi-generator/src/main/resources/go-experimental/api.mustache (hunk 2 above)
Comment (sebastien-rosset) [reply modules/openapi-generator/src/main/resources/go-experimental/api.mustache]: This is totally independent of the embedded go struct enhancement, maybe I should have raised a separate PR, but it takes time to manage multiple PRs.
Without this change, there can be compilation errors and static analysis issues. This is because the block (which is now in curly braces) is repeated multiple times depending on how many query parameters there are in the OAS. 

The issue is that this line will be generated multiple times:
```
t:={{paramName}}
```

**Comment by sebastien-rosset** [comment]
> > Using Golang embedded structs have zero impact on the JSON payload (vs a flattened structure). The payload remains exactly the same.
> 
> Yeah, but the serialized children will now contain parent attributes, which they didn't before. So the payload is going to be different, is it not?

Hmm, do you have an example, maybe I am missing something? Previously, with the following OAS schema:
Parent:
  property p1
  property p2

Child:
  property p3

The generated code for the child was a flattened struct that had p1, p2, p3. Hence in the JSON document these three properties were present.
Now "Parent" is embedded in the child; when the struct is serialized, it will still have the same p1, p2, p3 properties.

**Comment by sebastien-rosset** [comment]
Darn, the rebase from upstream didn't work as I expected. Let me fix this.

**Comment by bkabrda** [comment]
@sebastien-rosset I was looking wrong in terms of my above comments about different JSON, please disregard that.

In terms of changing the non-experimental go client, I think I was looking at an older version of your PR that has contained the same changes for non-experimental and experimental. With the current version of PR I'm +1 (with my Go language committee member hat on, can't speak for the changes to the core files).

**Comment by sebastien-rosset** [comment]
Sorry, I am still trying to figure out why changes are showing up for the following files, they shouldn't.
```
modules/openapi-generator-cli/src/main/java/org/openapitools/codegen/cmd/ConfigHelp.java
modules/openapi-generator/src/main/resources/JavaJaxRS/spec/libraries/quarkus/Dockerfile.jvm.mustache
modules/openapi-generator/src/main/resources/JavaJaxRS/spec/libraries/quarkus/application.properties.mustache
modules/openapi-generator/src/main/resources/JavaJaxRS/spec/libraries/quarkus/pom.mustache
pom.xml
```

**Comment by sebastien-rosset** [comment]
> @sebastien-rosset I was looking wrong in terms of my above comments about different JSON, please disregard that.
> 
> In terms of changing the non-experimental go client, I think I was looking at an older version of your PR that has contained the same changes for non-experimental and experimental. With the current version of PR I'm +1 (with my Go language committee member hat on, can't speak for the changes to the core files).

I have fixed the problem with rebase, the unrelated core changes are no longer in the PR.

**Comment by sebastien-rosset** [comment]
It appears there is an unrelated problem with Swift4 tests in master.

**Comment by sebastien-rosset** [comment]
@antihax, can you review this PR? thank you.

**Comment by sebastien-rosset** [comment]
> @sebastien-rosset I was looking wrong in terms of my above comments about different JSON, please disregard that.
> 
> In terms of changing the non-experimental go client, I think I was looking at an older version of your PR that has contained the same changes for non-experimental and experimental. With the current version of PR I'm +1 (with my Go language committee member hat on, can't speak for the changes to the core files).

@bkabrda , are you still ok with this fix? One big problem IMO is that the golang code does not compile. I tried to make it more clear in the PR summary. It would be great to fix this issue. thank you.

File: samples/client/petstore/go-experimental/go-petstore/model_dog.go [hunk 3]
```diff
@@ -16,59 +16,10 @@ import (
 
 // Dog struct for Dog
 type Dog struct {
-	ClassName string `json:"className"`
-	Color *string `json:"color,omitempty"`
+	Animal
```
Comment (jimschubert) [inline samples/client/petstore/go-experimental/go-petstore/model_dog.go]: This is very cool. I'm not a pro in Go by any means, but I've just learned about field promotion in anonymous nested structs. ;)
